File: audience/src/main/java/br/com/concretesolutions/audience/core/actor/ActivityChoreography.java
package br.com.concretesolutions.audience.core.actor;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

import br.com.concretesolutions.audience.core.Director;

final class ActivityChoreography implements Application.ActivityLifecycleCallbacks {

    @Override
    public void onActivityCreated(Activity activity, Bundle savedInstanceState) {

        // Configuration change finished
        Director.setInConfigurationChange(false);

        // Not an actor Activity
        if (!(activity instanceof Actor)) {
            return;
        }

        Director.getActorRegistry()
                .enroll((Actor) activity, savedInstanceState);
    }

    @Override
    public void onActivityStarted(Activity activity) {
    }

    @Override
    public void onActivityResumed(Activity activity) {
    }

    @Override
    public void onActivityPaused(Activity activity) {
    }

    @Override
    public void onActivityStopped(Activity activity) {
    }

    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle outState) {

        if (!(activity instanceof Actor))
            return;

        if (Director.isInConfigurationChange()) {
            Director.getActorRegistry()
                    .take5((Actor) activity, outState);
        }
    }

    @Override
    public void onActivityDestroyed(Activity activity) {

        if (!(activity instanceof Actor))
            return;

        Director.play().stop(Director.actorRef((Actor) activity));
    }
}

File: audience-core/src/main/kotlin/io/github/victorolinasc/audience/core/actor/ActivityChoreography.kt
package io.github.victorolinasc.audience.core.actor

import android.app.Activity
import android.app.Application
import android.os.Bundle
import android.support.v4.app.FragmentActivity

import io.github.victorolinasc.audience.core.Director

internal class ActivityChoreography : Application.ActivityLifecycleCallbacks {

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {

        // Configuration change finished
        Director.system.isInConfigurationChange = false

        if (activity is FragmentActivity) {
            activity
                    .supportFragmentManager
                    .registerFragmentLifecycleCallbacks(FragmentChoreography(), true)
        }

        // Not an actor Activity
        if (activity !is Actor) {
            return
        }

        Director.system.actorRegistry.enroll(activity as Actor, savedInstanceState)
    }

    override fun onActivityStarted(activity: Activity) {}

    override fun onActivityResumed(activity: Activity) {}

    override fun onActivityPaused(activity: Activity) {}

    override fun onActivityStopped(activity: Activity) {}

    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {

        if (activity !is Actor)
            return

        if (Director.system.isInConfigurationChange) {
            Director.system.actorRegistry
                    .take5(activity as Actor, outState)
        }
    }

    override fun onActivityDestroyed(activity: Activity) {

        if (activity !is Actor)
            return

        Director.system.stop(Director.actorRef(activity as Actor))
    }
}
